numba/experimental/jitclass/overloads.py

```diff
@@ -0,0 +1,169 @@
+"""
+Overloads for ClassInstanceType for built-in functions that call dunder methods
+on an object.
+"""
+import operator
+import sys
+
+from numba.core.extending import overload
+from numba.core.types import ClassInstanceType
+
+
+def _get_args(nargs=1):
+    return list("xyzabcdefg")[:nargs]
+
+
+def _simple_template(*attrs, nargs=1):
+    args = _get_args(nargs=nargs)
+    arg0 = args[0]
+
+    template = f"""
+def handler({",".join(args)}):
+    if not isinstance({arg0}, ClassInstanceType):
+        return
+"""
+    for attr in attrs:
+        assert isinstance(attr, str)
+        template += f"""
+    if "__{attr}__" in {arg0}.jit_methods:
+        return lambda {",".join(args)}: {arg0}.__{attr}__({",".join(args[1:])})
+"""
+    return template
+
+
+def _register_overload(func, template, glbls=None):
+    """
+    Register overload handler for func calling class attribute attr.
+    """
+    namespace = dict(ClassInstanceType=ClassInstanceType)
+    if glbls:
+        namespace.update(glbls)
+
+    exec(template, namespace)
+    handler = namespace["handler"]
+    overload(func)(handler)
+
+
+def register_simple_overload(func, *attrs, nargs=1):
+    template = _simple_template(*attrs, nargs=nargs)
+    _register_overload(func, template)
+
+
+@overload(bool)
+def class_bool(x):
+    if not isinstance(x, ClassInstanceType):
+        return
+
+    if "__bool__" in x.jit_methods:
+        return lambda x: x.__bool__()
+
+    if "__len__" in x.jit_methods:
+        return lambda x: x.__len__() != 0
+
+    return lambda x: True
```

Comment:
Is there some way to provide this as a default implementation of `bool` for all objects?

Reply:
I believe that
```
@overload(bool)
def fallback_bool(x):
    return lambda x: True
```
would not be a good idea (if I'm mistaken then cool).  We can add overloads on a type by type basis (e.g. https://github.com/numba/numba/pull/4715)